xapi-stdext-std: remove most functions, optimize escaping functions

psafont opened the pull request:
The Listext module has a lot of baggage that can be replace with Stdlib, Astring. And what cannot be replaced, it can be made better, especially the escaping.

There are quite a few changes, so it's better to review commit-by-commit.


I need to do some testing to make sure all changes here are safe and undraft this

psafont committed:
xapi-stdext-std: Make String.split's ~limit not optional

Users can instead opt for Astring.String's or stdlib's functions.

Signed-off-by: Pau Ruiz Safont <pau.ruizsafont@cloud.com>
Signed-off-by: Pau Ruiz Safont <pau.safont@vates.tech>

psafont committed:
xapi-stdext-std: Do not include String functions in XStringext

Also remove all indiscriminate opens against it

Signed-off-by: Pau Ruiz Safont <pau.ruizsafont@cloud.com>
Signed-off-by: Pau Ruiz Safont <pau.safont@vates.tech>

psafont committed:
xapi-stdext-std: Delete unused xstringext methods

Signed-off-by: Pau Ruiz Safont <pau.ruizsafont@cloud.com>
Signed-off-by: Pau Ruiz Safont <pau.safont@vates.tech>

psafont committed:
xapi-stdext-std: Replace String.has_substr with Astring's is_infix

Not only it's more efficient, but it's also more ergonomic

Signed-off-by: Pau Ruiz Safont <pau.ruizsafont@cloud.com>
Signed-off-by: Pau Ruiz Safont <pau.safont@vates.tech>

psafont committed:
xapi-stdext-std: replace String.filter_chars with Astring.filter

The former didn't have any tests and the performance is unknown

Signed-off-by: Pau Ruiz Safont <pau.ruizsafont@cloud.com>
Signed-off-by: Pau Ruiz Safont <pau.safont@vates.tech>

psafont committed:
xapi-stdext-std: replace String.split_f with Astring functions

This also allows to drop String.isspace

Signed-off-by: Pau Ruiz Safont <pau.ruizsafont@cloud.com>
Signed-off-by: Pau Ruiz Safont <pau.safont@vates.tech>

psafont committed:
xapi-stdext-std: change String.replace to replace characters

The few users that needed to replace strings, have been replaced with Astring's
cuts, as most of them were already segmenting strings, or they are run in very
specific, infrequent codepaths for efficiency to not matter.

Others have been replaced by Astring's filter as they were removing characters,
and the rest have been converted to the new String.replace.

map_unlikely can be removed from the interface and only have String.replaced
and String.replace

Signed-off-by: Pau Ruiz Safont <pau.ruizsafont@cloud.com>
Signed-off-by: Pau Ruiz Safont <pau.safont@vates.tech>

psafont committed:
xapi-stdext-std: Remove String.escaped, add String.replaced

String.replaced is an alias of map_unlikely. This names makes the intent
of the function clearer. Because a function to replace the characters is
exposed, users are less likely to fall into the pitfall of using lists.

Lists not only are very slow, but allow users to have more than one
replacement rule per character, possibly introducing mistakes.

If a plain match function cannot be produced and a list needs to be used,
users can convert it to a Char.Map and do the match with a find_opt.
This approach ends up being ~60-70% faster than using plain lists.

The benchmark comparing the new approach with the old one:

  String size 100:
    Optimized: 236.556 μs
    Reference: 1861.600 μs
    Improvement: 87.3% faster

  String size 500:
    Optimized: 1099.030 μs
    Reference: 9665.405 μs
    Improvement: 88.6% faster

  String size 1000:
    Optimized: 2198.777 μs
    Reference: 19115.019 μs
    Improvement: 88.5% faster

Signed-off-by: Pau Ruiz Safont <pau.safont@vates.tech>

cplaursen commented:
I'll run some system tests on this and have a proper look throughout the day.

psafont commented:
Yes, from the message of the commit that introduces it (https://github.com/xapi-project/xen-api/pull/6795/changes/1174f3a86f48512adac5a0fffe02725016a82ce3):
```
If a plain match function cannot be produced and a list needs to be used,
users can convert it to a Char.Map and do the match with a find_opt.
This approach ends up being ~60-70% faster than using plain lists.

The benchmark comparing the new approach with the old one:

  String size 100:
    Optimized: 236.556 μs
    Reference: 1861.600 μs
    Improvement: 87.3% faster

  String size 500:
    Optimized: 1099.030 μs
    Reference: 9665.405 μs
    Improvement: 88.6% faster

  String size 1000:
    Optimized: 2198.777 μs
    Reference: 19115.019 μs
    Improvement: 88.5% faster
```

It can also be easily run with `dune exec ocaml/libs/xapi-stdext/lib/xapi-stdext-std/bench/bench_xstringext.exe`

psafont committed:
xapi-stdext-std: test String.replace

I added some cases in preparation to change its implementation

Signed-off-by: Pau Ruiz Safont <pau.safont@vates.tech>

lindig commented:
Perfect example where some AI review would be helpful.

lindig commented:
Why is this targeting the LCM branch? You want to forward-port it afterwards?

psafont commented:
> Why is this targeting the LCM branch? You want to forward-port it afterwards?

I'll rebase it on master and do a backport. I didn't have time to do this after testing before I left for holidays

psafont commented:
The checks are not running, I'll try closing and opening the PR again, becauee I can't see any other way of triggering this

last-genius commented:
The first commit's title probably shouldn't be "String.split's ~limit _is not_ optional", but "_make_ String.split's limit not optional" or something along these lines?

psafont commented:
> The first commit's title probably shouldn't be "String.split's ~limit is not optional", but "make String.split's limit not optional" or something along these lines?

Thanks, I've reworded it
